termux-exec.c

```diff
@@ -7,232 +15,584 @@
+  if (getenv("TERMUX_EXEC_OPTOUT") != NULL) {
+    return execve_syscall(executable_path, argv, envp);
+  }
+
+  const bool termux_exec_debug = getenv("TERMUX_EXEC_DEBUG") != NULL;
+  if (termux_exec_debug) {
+    fprintf(stderr, LOG_PREFIX "Intercepting execve('%s'):\n", executable_path);
+    int tmp_argv_count = 0;
+    while (argv[tmp_argv_count] != NULL) {
+      fprintf(stderr, LOG_PREFIX "   argv[%d] = '%s'\n", tmp_argv_count, argv[tmp_argv_count]);
+      tmp_argv_count++;
+    }
+  }
+
+  // Size of TERMUX_BASE_DIR_MAX_LEN + "/usr/bin/sh".
+  char termux_base_dir_buf[TERMUX_BASE_DIR_MAX_LEN + 16];
+
+  const char *termux_base_dir = get_termux_base_dir(termux_base_dir_buf);
+  if (termux_exec_debug) {
+    fprintf(stderr, LOG_PREFIX "Using base dir: '%s'\n", termux_base_dir);
+  }
+
+  const char *orig_executable_path = executable_path;
+
+  char executable_path_buffer[PATH_MAX];
+  executable_path = termux_rewrite_executable(termux_base_dir, executable_path, executable_path_buffer,
+                                              sizeof(executable_path_buffer));
+  if (termux_exec_debug && executable_path_buffer == executable_path) {
+    fprintf(stderr, LOG_PREFIX "Rewritten path: '%s'\n", executable_path);
+  }
+
+  if (access(executable_path, X_OK) != 0) {
+    // Error out if the file is not executable:
+    errno = EACCES;
+    return -1;
+  }
+
+  int fd = open(executable_path, O_RDONLY);
+  if (fd == -1) {
+    errno = ENOENT;
+    return -1;
+  }
+
+  // execve(2): "The kernel imposes a maximum length on the text that follows the "#!" characters
+  // at the start of a script; characters beyond the limit are ignored. Before Linux 5.1, the
+  // limit is 127 characters. Since Linux 5.1, the limit is 255 characters."
+  // We use one more byte since inspect_file_header() will null terminate the buffer.
+  char header[256];
+  ssize_t read_bytes = read(fd, header, sizeof(header) - 1);
+  close(fd);
+
+  struct file_header_info info = {
+      .interpreter = NULL,
+      .interpreter_arg = NULL,
+  };
+  inspect_file_header(termux_base_dir, header, read_bytes, &info);
+
+  if (!info.is_elf && info.interpreter == NULL) {
+    if (termux_exec_debug) {
+      fprintf(stderr, LOG_PREFIX "Not ELF or shebang, returning ENOEXEC\n");
+    }
+    errno = ENOEXEC;
+    return -1;
+  }
+
+  if (info.interpreter != NULL) {
+    executable_path = info.interpreter;
+    if (termux_exec_debug) {
+      fprintf(stderr, LOG_PREFIX "Path to interpreter from shebang: '%s'\n", info.interpreter);
+    }
+  }
+
+  char normalized_path_buffer[PATH_MAX];
+  executable_path = normalize_path(executable_path, normalized_path_buffer);
+
+  char **new_allocated_envp = NULL;
+
+  const char **new_argv = NULL;
+
+  // Only wrap with linker:
+  // - If running on an Android 10+ where it's necesssary, and
+  // - If trying to execute a file under the termux base directory
+  bool wrap_in_linker =
+#ifdef __ANDROID__
+      android_get_device_api_level() >= 29 &&
```

Comment:
should this be determined at compile time via `__ANDROID_API__` rather than at runtime? i thought the app's target api is the issue, not the device it's running on?

Reply:
If `targetSdkVersion` is `>= 29` of built app and `sdkVersion` is `>= 29` of device it is running on, only then restrictions engage.

Env variables checks need to be added here, one for `wrap_in_linker = ... && ... $TERMUX__TARGET_SDK >= 29` (the currently engaged target sdk version extract from `ApplicationInfo.seInfo` (check termux settings about), different from `$TERMUX_APP__TARGET_SDK` which is what termux app is built with), since if a `sharedUserId` plugin app is installed with a lower target sdk or if running non-playstore builds with older target sdk, then android will use lowest sdk for all apps after reboot and we should not wrap with linker in that case. Currently, only `$TERMUX_APP__TARGET_SDK` is exported, need to export `TERMUX__TARGET_SDK`. Not everyone will be willing to live with all the issues of linker solution, even if we bump `targetSdkVersion` for termux-app for all release sources and users may install a plugin with a lower target sdk to run in legacy mode.

https://matrix.to/#/!rrHhKvLYtBVDOISphu:gitter.im/$zl7im-r7aR5jpxpzbpMJgb9xwWzKW69mapNPmSLyMNU?via=gitter.im&via=matrix.termux.club&via=matrix.org

https://github.com/termux/termux-app/blob/e2f0edf4d20bc17c99b39c551708296f2c5c7939/termux-shared/src/main/java/com/termux/shared/termux/shell/command/environment/TermuxAppShellEnvironment.java#L47

https://github.com/termux/termux-app/issues/2445#issuecomment-985484952

A second check needs to be added for `wrap_in_linker = ... && ... $TERMUX_SE_POLICY__APP_DATA_FILE_EXECUTE_ALLOW != "true"` to prevent linker wrap for if app (via in-app support in next version) or user themselves or custom ROM has disabled selinux restrictions with root on `sdkVersion` is `>= 29`, normally the variable will not be set.

Reply:
Another issue here is that if termux app is installed as a system app with `system_app`, `priv_app` or `platform_app` process context, then restrictions would engage for older target sdk and device sdk versions too, so instead of `$TERMUX__CUR_TARGET_SDK` check, we may need to check if `/proc/self/attr/current` or `$TERMUX__SE_PROCESS_CONTEXT` does not start with `u:r:untrusted_app_25:` and does not start with `u:r:untrusted_app_27:` to wrap with linker.

Reply:
> If `targetSdkVersion` is `>= 29` of built app and `sdkVersion` is `>= 29` of device it is running on, only then restrictions engage.

So check should be changed to `android_get_application_target_sdk_version() > 29` here

Reply:
Thanks!

It's updated to `android_get_device_api_level() >= 29 && android_get_application_target_sdk_version() >= 29` now!